core/src/main/java/org/elasticsearch/index/query/ExistsQueryBuilder.java

```diff
@@ -143,18 +149,49 @@ public static Query newFilter(QueryShardContext context, String fieldPattern) {
         }
 
         if (fields.size() == 1) {
-            Query filter = fieldNamesFieldType.termQuery(fields.iterator().next(), context);
+            String field = fields.iterator().next();
+            MappedFieldType fieldType = context.getMapperService().fullName(field);
+            if (fieldType == null) {
+                // The field does not exist as a leaf but could be an object so check for an object mapper
+                if (context.getObjectMapper(field) != null) {
+                    return newObjectFieldExistsQuery(context, field);
+                }
+                return Queries.newMatchNoDocsQuery("No field \"" + field + "\" exists in mappings.");
+            }
+            Query filter = fieldType.existsQuery(context);
             return new ConstantScoreQuery(filter);
         }
 
         BooleanQuery.Builder boolFilterBuilder = new BooleanQuery.Builder();
         for (String field : fields) {
-            Query filter = fieldNamesFieldType.termQuery(field, context);
-            boolFilterBuilder.add(filter, BooleanClause.Occur.SHOULD);
+            MappedFieldType fieldType = context.getMapperService().fullName(field);
+            if (fieldType == null) {
+                // The field does not exist as a leaf but could be an object so check for an object mapper
+                if (context.getObjectMapper(field) != null) {
+                    boolFilterBuilder.add(newObjectFieldExistsQuery(context, field), BooleanClause.Occur.SHOULD);
+                }
+            } else {
+                Query filter = fieldType.existsQuery(context);
+                boolFilterBuilder.add(filter, BooleanClause.Occur.SHOULD);
+            }
         }
         return new ConstantScoreQuery(boolFilterBuilder.build());
     }
 
+    private static Query newObjectFieldExistsQuery(QueryShardContext context, String field) {
+        BooleanQuery.Builder booleanQuery = new BooleanQuery.Builder();
+        for (Iterator<Mapper> itr = context.getObjectMapper(field).iterator(); itr.hasNext();) {
+            Mapper childField = itr.next();
+            if (childField instanceof ObjectMapper) {
+                booleanQuery.add(newObjectFieldExistsQuery(context, childField.name()), Occur.SHOULD);
+            } else {
+                Query existsQuery = context.getMapperService().fullName(childField.name()).existsQuery(context);
+                booleanQuery.add(existsQuery, Occur.SHOULD);
+            }
+        }
+        return new ConstantScoreQuery(booleanQuery.build());
+    }
```

Comment:
Now that I think more about it, I think this impl has potential problems with multiple types, since getObjectMapper will return the mapper of one of the types, meaning it might miss sub fields if they only exist in another type (those mapper classes are per-type on the contrary to the field type classes). So maybe we should resolve fields that match field + ".*" and build a bool query out of it (and no need for recursion since I would expect the glob to match fields of sub objects too)

Reply:
fixed